### PR TITLE
[non-strict][leaf_function] make `leaf_function` decorator work in non-strict

### PR DESCRIPTION
**Pull request by yushangdi**
As title. 

Add a python wrapper so we capture the HOP in non-strict.

Tested the following cases:

- `@leaf_function` with make_fx tracing, only forward
- `@leaf_function` with aot_function, both forward and backward 
- `@leaf_function` for closure, data-dependent, pytree inputs, module inputs
- `@leaf_function` properly error when there is no fake impl. or escaped gradients
- `@leaf_function` works with the invoke_subgraph backend that mix make_fx and torch.compile

```
python test/dynamo/test_decorators.py -k leaf_function
python test/functorch/test_leaf_function.py 
```
 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #174636
* __->__ #174635

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

### CODE REVIEW

File: torch/_dynamo/variables/torch.py [hunk 1]
```diff
@@ -2695,20 +2697,6 @@ def is_module_variable(
             kwargs_var = VariableTracker.build(tx, kwargs)
             return args_var, kwargs_var
 
-        def convert_modules_to_states(
```
Comment (yushangdi) [inline torch/_dynamo/variables/torch.py:2696]: this is just moved to invoke_leaf_function so we can re-use it in the python wrapper

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 2]
```diff
@@ -38,6 +38,26 @@ class LeafModuleState(NamedTuple):
     named_buffers: dict[str, torch.Tensor]
 
 
+def convert_modules_to_states(values: Any, module_to_index: dict[int, int]) -> Any:
```
Comment (yushangdi) [inline torch/_higher_order_ops/invoke_leaf_function.py:71]: this is just moved from torch.py

File: torch/_dynamo/decorators.py [hunk 3]
```diff
@@ -520,6 +570,23 @@ def my_leaf_fn_fake(x):
 
     @functools.wraps(fn)
     def inner(*args: _P.args, **kwargs: _P.kwargs) -> _R:
+        if get_proxy_mode() is not None or detect_fake_mode(args) is not None:
```
Comment (ydwu4) [inline torch/_dynamo/decorators.py]: Could we do not torch.compiler.is_dynamo_compiling?

File: test/functorch/test_leaf_function.py [hunk 4]
```diff
@@ -0,0 +1,673 @@
+# Owner(s): ["oncall: pt2"]
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py:1]: Don't have to happen in this PR. we probably wants to consolidate all leaf_function related tests in a single file to make it easier to test the changes.

File: torch/_dynamo/decorators.py [hunk 5]
```diff
@@ -259,6 +263,52 @@ def deregister() -> None:
...
+    )[0]:
+        if isinstance(val, torch.nn.Module) and id(val) not in module_to_index:
+            module_to_index[id(val)] = len(captured_modules)
+            captured_modules.append(val)
+
+    processed = convert_modules_to_states((args, kwargs), module_to_index)
+    flat_args, input_spec = pytree.tree_flatten(processed)
+
+    def _make_wrapper(impl: Callable[..., Any]) -> Callable[..., Any]:
+        def wrapper(*flat_args: Any) -> Any:
```
Comment (ydwu4) [inline torch/_dynamo/decorators.py]: is it possible to share the wrapper with the compiled path to reduce divergence?

**Review by ydwu4 — COMMENTED** [review]
Looks good. left some comments.

**Review by ydwu4 — APPROVED** [review]

File: test/functorch/test_leaf_function.py (hunk 4 above)
Comment (yushangdi) [reply test/functorch/test_leaf_function.py:1]: sure. I can make a follow up PR to consolidate into `test/higher_order_ops/`

**Review by SherlockNoMad — APPROVED** [review]

File: torch/_dynamo/decorators.py (hunk 3 above)
Comment (yushangdi) [reply torch/_dynamo/decorators.py]: if we use `not torch.compiler.is_dynamo_compiling()`, then the eager code will go through `_invoke_leaf_function_python` instead of directly calling `fn(*args, **kwargs)` . I think ideally we want to interfere with eager code less?

File: torch/_dynamo/decorators.py (hunk 5 above)
Comment (yushangdi) [reply torch/_dynamo/decorators.py]: sure, shared now!

File: torch/_dynamo/decorators.py (hunk 3 above)
Comment (ydwu4) [reply torch/_dynamo/decorators.py]: I don't like the get_proxy_mode() detect_fake_mode(args) lol because they're relying on the implementation instead of the interface. 

Btw, does the first pass of aot_autograd (the run functionalization and collect meta pass) trigger real execution? It would be wrong if it does: the contract has been real_fn should only executed once per call and not called during compilation. Can add a test to verify?

If it's indeed called multiple times. Here are the possible alternatives:

1. Call _invoke_leaf_function_python in eager. In this case we correctly handle the issue that real_fn can get executed multiple times. Even though this means some eager overhead. 

2. We'll always torch.compile even in eager mode like cond/flex_attention. In this case, we avoid the python wrapper in invoke_leaf_function and consolidate on a single path but this might against the no torch.compile stand in the way principle. 

I'm leaning towards 1, any thoughts?

File: torch/_higher_order_ops/utils.py [hunk 6]
```diff
@@ -599,14 +599,20 @@ def unmask_none_gradients(grads, operands):
...
 def _maybe_fake_prop_ignore_unbacked(fn, args):
-    with ExitStack() as ctx_stack:
-        if (fake_mode := detect_fake_mode(args)) is not None:
-            ctx_stack.enter_context(fake_mode)
-            if fake_mode.shape_env is not None:
-                ctx_stack.enter_context(
-                    fake_mode.shape_env.ignore_fresh_unbacked_symbols()
-                )
-        return fn(*args)
+    with suspend_functionalization(), disable_functional_mode():
```
Comment (yushangdi) [inline torch/_higher_order_ops/utils.py:602]: disable functionalization and proxy mode, similar to what we do for `materialize_as_graph` in `gen_schema`

**Review by ydwu4 — APPROVED** [review]
